src/shapirowilk.jl

```diff
@@ -0,0 +1,248 @@
+
+function pvalue(t::ShapiroWilkTest)
+    n = length(t.SWc)
+    W = t.W
+
+    if iszero(censored_ratio(t))
+        if n == 3 # exact by Shapiro&Wilk 1965
+            # equivalent to 6/π * (asin(sqrt(W)) - asin(sqrt(3/4)))
+            return 1 - 6acos(sqrt(W)) / π
+        elseif n ≤ 11 # Royston 1992
+            γ = __RS92_G(n)
+            if γ ≤ log1p(-W)
+                return zero(W)
+            end
+            w = -log(γ - log1p(-W))
+            μ = __RS92_C3(n)
+            σ = exp(__RS92_C4(n))
+        elseif 12 ≤ n # Royston 1992
+            w = log1p(-W)
+            μ = __RS92_C5(log(n))
+            σ = exp(__RS92_C6(log(n)))
+        end
+        return ccdf(Normal(μ, σ), w)
+    else
+        throw("censored samples not implemented yet")
+        # to implement censored samples follow Royston 1993 Section 3.3
+    end
+end
+
+"""
+    ShapiroWilkTest(X::AbstractArray{<:Real}, SWc::SWCoeffs=SWCoeffs(length(X)); kwargs...)
+Perform a Shapiro-Wilk test of normality on `X`.
+
+This julia implementation is based the method of Royston (1992).
+The calculation of the p-value is exact for `N = 3`, and for ranges
+`4 ≤ N ≤ 11` and `12 ≤ N ≤ 5000` (Royston 1992) two separate approximations
+for p-values are used.
+
+Implements: [`pvalue`](@ref)
+
+# Notes (Royston 1993)
+* While the (approximated) W-statistic will be accurate for large sample size
+  (`N > 2000`), returned p-values may not be reliable.
+* Censoring too much data (`(N - N1) / N > 0.8`, where `N1` is (upper)
+  uncensored data length), or when the sample size is small (`N < 20`) may
+  produce unreliable p-values.
+
+# Implementation notes
+* The current implementation DOES NOT implement p-values for censored data.
+* If multiple Shapiro-Wilk tests are to be performed on samples of same
+  cardinality it is beneficial to pass `SWc` for re-use.
+* For maximal performance sorted `X` should be passed and indicated with
+  `sample_sorted=true` keyword argument.
+
+# References
+Shapiro, S. S., & Wilk, M. B. (1965). An Analysis of Variance Test for Normality
+(Complete Samples). *Biometrika*, 52, 591–611.
+[doi:10.1093/BIOMET/52.3-4.591](https://doi.org/10.1093/BIOMET/52.3-4.591).
+
+Royston, P. (1992). Approximating the Shapiro-Wilk W-test for non-normality.
+*Statistics and Computing*, 2(3), 117–119.
+[doi:10.1007/BF01891203](https://doi.org/10.1007/BF01891203)
+
+Royston, P. (1993). A Toolkit for Testing for Non-Normality in Complete and
+Censored Samples. Journal of the Royal Statistical Society Series D
+(The Statistician), 42(1), 37–43.
+[doi:10.2307/2348109](https://doi.org/10.2307/2348109)
+
+Royston, P. (1995). Remark AS R94: A Remark on Algorithm AS 181: The W-test for
+Normality. *Journal of the Royal Statistical Society Series C
+(Applied Statistics)*, 44(4), 547–551.
+[doi:10.2307/2986146](https://doi.org/10.2307/2986146).
+"""
+function ShapiroWilkTest(
+    sample::AbstractArray{<:Real},
+    SWc::SWCoeffs=SWCoeffs(length(sample));
+    N1=length(sample),
```

Comment:
`N1` isn't a super explicit name. Maybe use something else?

Reply:
I settled on `lower_uncensored` which seems descriptive and accurate

Reply:
Maybe it would be more intuitive to specify the number of censored observations? "uncensored" sounds a bit funny. In typical use, which of the two quantities people are likely to store, number of censored or uncensored?

And maybe "lower" doesn't need to be in the argument name, unless you plan to add another one for "upper"?

Reply:
we could just make it into `censored = 0`;
I used `lower_uncensored` and `uncensored_ratio` to be consistent with the language of Royston 1992, but we can shape the user interface as we like ;)

Reply:
OK. Is there any reason to censor large observations rather than small ones? Intuitively and without knowing anything about this, I would have censored observations with the largest squared deviation from the mean.

Reply:
Bump.